File: Algorithms/Java/Intersection of Two Linked Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
    	ListNode nodeA = headA;
		ListNode nodeB = headB;
		int lenA = 0, lenB = 0;
		while (nodeA != null) {
			nodeA = nodeA.next;
			lenA++;
		}
		while (nodeB != null) {
			nodeB = nodeB.next;
			lenB++;
		}
		nodeA = headA;
		nodeB = headB;
		if (lenA > lenB) {
			for (int i = 0; i < lenA - lenB; i++)
				nodeA = nodeA.next;
		} else if (lenA < lenB) {
			for (int i = 0; i < lenB - lenA; i++)
				nodeB = nodeB.next;
		}
		while (nodeA != nodeB) {
			nodeA = nodeA.next;
			nodeB = nodeB.next;
		}
		return nodeA;
    }
}

File: Algorithms/Python/Intersection-of-Two-Linked-Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):

    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        numA, numB = 0, 0
        nodeA, nodeB = headA, headB
        while nodeA:
            numA += 1
            nodeA = nodeA.next
        while nodeB:
            numB += 1
            nodeB = nodeB.next
        nodeA, nodeB = headA, headB
        if numA > numB:
            for i in range(numA - numB):
                nodeA = nodeA.next
        else:
            for i in range(numB - numA):
                nodeB = nodeB.next
        while nodeA != nodeB:
            nodeA, nodeB = nodeA.next, nodeB.next
        return nodeA
